tools/autograd/templates/Functions.cpp

```diff
@@ -972,22 +972,49 @@ Tensor softplus_double_backward(const Tensor & grad, const Tensor & input, Scala
 //
 // Finally, we note that these general operations are expensive, so we apply the
 // following optimizations:
-//   Add step (0): For all output dimension `d` with output stride 0, sum the
-//                 gradients along dimension `d` (don't keepdim), and remove
-//                 dimension `d` from output size and stride.
-//                 (An optimization for "expand" cases so we may avoid step (3))
-//  Only apply step (3) when input tensor has overlapping memory.
+//   Step (3): only apply the procedure outlined above when input tensor has
+//             overlapping memory.
+//   Add step (0):
+//     (a): [optimize for "expand" cases so we may avoid index_add in step (2)]
+//          For all output dimension `d` with output stride 0, sum the gradients
+//          along dimension `d` (don't keepdim), and remove dimension `d` from
+//          output size and stride.
+//     (b): [optimize for "permute" (and "transpose") cases so we may avoid
+//           creating "storage" tensor]
+//          (i):  Sort the strides of input and output geometries. We need to do
+//                this anyways to detech overlapping memory.
+//                See NOTE [ Detecting Memory Overlap Within A Strided Tensor ].
+//          (ii): Check if they can be matched via some permutation. If so,
+//                return `grad` permuted in reverse order.
```

Comment:
Err, you're not applying the permutation that reverses the order of the strides, you're applying the inverse permutation, right?